[FSDP()][10/N][11/N] Introduce composable (ctor only)

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88123 [FSDP] Rename `unflat_param_name` -> `fqn` for consistency
* #88122 [FSDP] Simplify `_get_buffer_names()`
* #88121 [FSDP] Remove unneeded `torch.no_grad()` context when offloading to CPU
* #88120 [FSDP][Docs] Add note mentioning rate limiter for backward prefetch
* #88040 [FSDP()][27/N] Add forward hook registration
* #87941 [FSDP()][26/N] Move `_lazy_init()` into `_fsdp_root_pre_forward()`
* #87940 [FSDP()][25/N] Add `_post_forward_reshard()`
* #87939 [FSDP()][24/N] Refactor `_lazy_init()`
* #87938 [FSDP()][23/N] Refactor handle attr initialization
* #87937 [FSDP] Simplify `_reset_lazy_init()`
* #87936 [FSDP()][22/N] Refactor `_cast_buffers()` in `_lazy_init()`
* #87935 [FSDP()][21/N] Refactor `_buffer_name_to_orig_dtype` computation
* #87934 [FSDP] Rename `dtype` to `buffer_name_to_dtype`
* #87933 [FSDP] Remove `device` arg from `_cast_buffers()`
* #87932 [FSDP()][20/N][Easy] Move functions in file
* #87931 [FSDP()][18/N] Refactor `pre_forward_unshard()`
* #87930 [FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`
* #87929 [FSDP()][16/N] Refactor post-forward/pre-backward
* #87928 [FSDP()][15/N] Refactor `_init_streams()`
* #87927 [FSDP()][14/N] Refactor pre-forward/post-backward
* #87926 [FSDP()][13/N] Refactor unshard/reshard/grads
* #87925 [FSDP()][12/N] Easy cleanup
* **#87924 [FSDP()][10/N][11/N] Introduce composable (ctor only)**
* #87923 [FSDP()][9/N] Refactor ctor (continued)

This PR introduces the composable FSDP API (with constructor semantics only) along with some further constructor refactoring. A notable contribution here is `_get_submodule_to_states()`, which performs auto wrapping without actually wrapping.

Commit:
[FSDP()][10/N][11/N] Introduce composable (ctor only)

[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

mrshenli commented on test/distributed/fsdp/test_composable_fsdp.py:
```diff
@@ -0,0 +1,215 @@
+# Owner(s): ["oncall: distributed"]
```
no need to change this PR. Let's move this to the new `_composable` folder

mrshenli commented on test/distributed/fsdp/test_composable_fsdp.py:
```diff
@@ -0,0 +1,215 @@
+            use_orig_params=True,
+        )
+        composable_module = copy.deepcopy(local_model)
+        fsdp_state = fully_sharded_data_parallel(
```
no need to change this PR either. Let's move to contract's state in follow up PRs.

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"




[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/_wrap_utils.py:
```diff
@@ -53,3 +70,96 @@ def _auto_wrap(
+    parameters were assigned to a parent submodule instead.
+    """
+    # Record the modules to wrap without actually wrapping
+    wrapped_modules: List[nn.Module] = []  # these are only logically wrapped
```
no need to change anything. JFYI, existing tracing techniques usually have issues on tracing inplace tensor functions and they have to do special-handling to workaround. There is a discussion whether we should provide full functional ops and can do another optimization pass to re-inplace ops. It might worth considering whether we would like to have similar functional design in our APIs as well.

mrshenli reviewed (approved):
Would I be correct that the main new thing in this PR is the `fuly_sharded_data_parallel` in `_fsdp.py`, and others are mainly refactoring so that we can call those state init in both wrapper and composable APIs?

awgu commented:
> Would I be correct that the main new thing in this PR is the `fuly_sharded_data_parallel` in `_fsdp.py`, and others are mainly refactoring so that we can call those state init in both wrapper and composable APIs?

Yes, exactly :)

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"


This PR introduces the composable FSDP API (with constructor semantics only) along with some further constructor refactoring. A notable contribution here is `_get_submodule_to_states()`, which performs auto wrapping without actually wrapping.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"


This PR introduces the composable FSDP API (with constructor semantics only) along with some further constructor refactoring. A notable contribution here is `_get_submodule_to_states()`, which performs auto wrapping without actually wrapping.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][10/N][11/N] Introduce composable (ctor only)"


This PR introduces the composable FSDP API (with constructor semantics only) along with some further constructor refactoring. A notable contribution here is `_get_submodule_to_states()`, which performs auto wrapping without actually wrapping.

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/_fsdp.py:
```diff
@@ -0,0 +1,69 @@
+from typing import Callable, cast, Iterable, Optional, Union
```
nit: assume you are going to move it to '_composable' folder?

awgu replied:
Yup! I will do that in follow-ups.

zhaojuanmao commented on torch/distributed/fsdp/_wrap_utils.py:
```diff
@@ -53,3 +70,96 @@ def _auto_wrap(
+    wrapped_modules.reverse()
+    wrapped_modules_set = set(wrapped_modules)
+    for submodule in wrapped_modules:
+        # Perform a BFS from `submodule` and record all unvisited state that is
```
nice algorithm!!
